keras_nlp/utils/text_generation.py

```diff
@@ -288,63 +298,91 @@ def token_probability_fn(inputs):
 
-    if prompt.shape[-1] == 0:
-        raise ValueError(
-            "Length of `prompt` is 0, please provide a non-empty `prompt`."
-        )
-
     input_is_1d = prompt.shape.rank == 1
     if input_is_1d:
         prompt = prompt[tf.newaxis, :]
-    _validate_token_probability_fn(token_probability_fn, prompt)
 
-    batch_size, length = prompt.shape
+    batch_size, length = _get_prompt_shape(prompt)
+    prompt, mask = _pad_prompt(prompt, max_length)
+
+    vocab_size = _validate_token_probability_fn(
+        token_probability_fn, prompt, return_vocab_size=True
+    )
+
     if length >= max_length:
         return tf.squeeze(prompt) if input_is_1d else prompt
+    # Initialize beam with shape `(batch_size, num_beams, length)`.
+    beams = tf.repeat(tf.expand_dims(prompt, axis=1), num_beams, axis=1)
+
+    # Initialize beams_prob with shape `(batch_size, num_beams)`
+    beams_prob = tf.zeros([batch_size, 1], dtype=tf.float32)
+    if num_beams > 1:
+        beams_prob = tf.concat(
+            [beams_prob, tf.fill((batch_size, num_beams - 1), -3.40282e38)],
+            axis=-1,
+        )
+
+    def one_step(beams, beams_prob, length):
+        truncated_beams = beams[..., :length]
+
+        # logits is of shape [batch_size, num_beams * vocab_size]
+        logits = tf.reshape(
+            token_probability_fn(
+                tf.reshape(truncated_beams, shape=[batch_size * num_beams, -1])
+            ),
+            shape=[batch_size, -1],
+        )
 
-    # Initialize beam.
-    beams = tf.expand_dims(prompt, 1)
-    beams_prob = tf.zeros([batch_size, 1])
-    i = length
-    while i < max_length:
-        beam_size = beams.shape[1]
-        beam_preds = []
-        for j in range(beam_size):
-            preds = token_probability_fn(beams[:, j, :])
-            if from_logits:
-                preds = keras.activations.softmax(preds, axis=-1)
-            beam_preds.append(preds)
-        stacked_preds = tf.stack(beam_preds, axis=1)
-        vocab_size = stacked_preds.shape[2]
-        logits = tf.reshape(stacked_preds, [batch_size, beam_size * vocab_size])
         probs = tf.math.log(logits) + tf.repeat(
             beams_prob, repeats=vocab_size, axis=1
         )
-        num_beams = min(beam_size * vocab_size, num_beams)
+
         candidate_prob, candidate_indexes = tf.math.top_k(
             probs, k=num_beams, sorted=False
         )
         candidate_beam_indexes = candidate_indexes // vocab_size
         next_token = candidate_indexes % vocab_size
 
         beams = tf.gather(beams, candidate_beam_indexes, axis=1, batch_dims=1)
-        beams = tf.concat([beams, next_token[..., tf.newaxis]], axis=-1)
+
+        next_token = tf.where(
+            condition=mask[..., length, tf.newaxis],
+            x=beams[..., length],
+            y=next_token,
+        )
+        beams = tf.tensor_scatter_nd_update(
+            tensor=beams,
+            indices=tf.concat(
+                [
+                    tf.cast(
```

Comment:
What is the overall shape/format you are trying to achieve here with `indices`? Feel this could be simplified, or at least made more readable.

Reply:
So, assume batch_size = 2, num_beams = 3, length = 2.

```
tf.cast(tf.where(tf.ones((batch_size, num_beams), tf.bool)), dtype=tf.int32)
```

will give us

```
<tf.Tensor: shape=(6, 2), dtype=int32, numpy=
array([[0, 0],
       [0, 1],
       [0, 2],
       [1, 0],
       [1, 1],
       [1, 2]], dtype=int32)>
```

We then concatenate every row with length (i.e., the index of the "column" to be updated):

```
<tf.Tensor: shape=(6, 3), dtype=int32, numpy=
array([[0, 0, 2],
       [0, 1, 2],
       [0, 2, 2],
       [1, 0, 2],
       [1, 1, 2],
       [1, 2, 2]], dtype=int32)>
```

So, essentially, iterating over all possible values of `(:batch_size, :num_beams)`, keeping the length fixed.

Reply:
Do you think moving this part `tf.cast(tf.where(tf.ones((batch_size, num_beams), tf.bool)), dtype=tf.int32)` out of the while loop is better? It remains the same in every iteration, anyway

Reply:
We can do the same for other text gen functions (move `tf.range(batch_size)` outside the loop): https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/utils/text_generation.py#L185

Reply:
I've split it into two...let me know if it's okay.